osu.Game/Screens/Play/HUD/SongProgress.cs

```diff
@@ -0,0 +1,90 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+#nullable disable
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using osu.Framework.Allocation;
+using osu.Framework.Bindables;
+using osu.Framework.Graphics.Containers;
+using osu.Framework.Timing;
+using osu.Game.Beatmaps;
+using osu.Game.Rulesets.Objects;
+using osu.Game.Rulesets.UI;
+using osu.Game.Skinning;
+
+namespace osu.Game.Screens.Play.HUD
+{
+    public abstract class SongProgress : OverlayContainer, ISkinnableDrawable
+    {
+        public bool UsesFixedAnchor { get; set; }
+
+        [Resolved(canBeNull: true)]
+        private GameplayClock gameplayClock { get; set; }
+
+        [Resolved(canBeNull: true)]
+        private GameplayClockContainer gameplayClockContainer { get; set; }
+
+        [Resolved(canBeNull: true)]
+        private DrawableRuleset drawableRuleset { get; set; }
+
+        [Resolved(canBeNull: true)]
+        private IBindable<WorkingBeatmap> beatmap { get; set; }
+
+        private IClock referenceClock;
+        private IEnumerable<HitObject> objects;
+
+        public IEnumerable<HitObject> Objects
+        {
+            set => UpdateObjects(objects = value);
+        }
+
+        protected double FirstHitTime => objects.FirstOrDefault()?.StartTime ?? 0;
+
+        //TODO: this isn't always correct (consider mania where a non-last object may last for longer than the last in the list).
+        protected double LastHitTime => objects.LastOrDefault()?.GetEndTime() ?? 0;
+
+        protected double FirstEventTime { get; private set; }
+
+        protected abstract void UpdateProgress(double progress, double time, bool isIntro);
```

Comment:
It's confusing to have `progress` and `time` together, since the former relies on frame stable time while the latter relies on gameplay clock time. `DefaultSongProgress` should just use `gameplayClock.CurrentTime` directly.